src/Rules/Library.hs

```diff
@@ -51,45 +81,59 @@ buildDynamicLib context@Context{..} = do
 
 buildPackageLibrary :: Context -> Rules ()
 buildPackageLibrary context@Context {..} = do
-    let libPrefix = "//" ++ contextDir context -/- "libHS" ++ pkgName package
-    libPrefix ++ "*" ++ (waySuffix way <.> "a") %%> \a -> do
+    root <- buildRootRules
+    pkgId <- case pkgCabalFile package of
+      Just file -> liftIO (parseCabalPkgId file)
+      Nothing   -> return (pkgName package)
+    let libPrefix = root -/- buildDir context -/- "libHS" ++ pkgId
+        archive = libPrefix ++ (waySuffix way <.> "a")
+    archive %%> \a -> do
         objs <- libraryObjects context
         asuf <- libsuf way
         let isLib0 = ("//*-0" ++ asuf) ?== a
         removeFile a
         if isLib0 then build $ target context (Ar Pack stage) []   [a] -- TODO: Scan for dlls
                   else build $ target context (Ar Pack stage) objs [a]
 
-        synopsis <- traverse pkgSynopsis (pkgCabalFile package)
+        synopsis <- pkgSynopsis context
         unless isLib0 . putSuccess $ renderLibrary
             (quote (pkgName package) ++ " (" ++ show stage ++ ", way "
             ++ show way ++ ").") a synopsis
 
+    library context
+
 buildPackageGhciLibrary :: Context -> Rules ()
 buildPackageGhciLibrary context@Context {..} = priority 2 $ do
-    let libPrefix = "//" ++ contextDir context -/- "HS" ++ pkgName package
-    libPrefix ++ "*" ++ (waySuffix way <.> "o") %> \obj -> do
+    root <- buildRootRules
+    pkgId <- case pkgCabalFile package of
+      Just file -> liftIO $ parseCabalPkgId file
+      Nothing   -> return (pkgName package)
+
+    let libPrefix = root -/- buildDir context -/- "HS" ++ pkgId
+        o = libPrefix ++ "*" ++ (waySuffix way <.> "o")
+    o %> \obj -> do
         objs <- allObjects context
         need objs
-        build $ target context Ld objs [obj]
+        build $ target context (Ld stage) objs [obj]
 
 allObjects :: Context -> Action [FilePath]
 allObjects context = (++) <$> nonHsObjects context <*> hsObjects context
 
 nonHsObjects :: Context -> Action [FilePath]
 nonHsObjects context = do
-    path    <- buildPath context
     cObjs   <- cObjects context
-    cmmSrcs <- pkgDataList (CmmSrcs path)
+    cmmSrcs <- interpretInContext context (getConfiguredCabalData ConfCabal.cmmSrcs)
     cmmObjs <- mapM (objectPath context) cmmSrcs
     eObjs   <- extraObjects context
     return $ cObjs ++ cmmObjs ++ eObjs
 
 cObjects :: Context -> Action [FilePath]
 cObjects context = do
-    path <- buildPath context
-    srcs <- pkgDataList (CSrcs path)
-    mapM (objectPath context) srcs
+    srcs <- interpretInContext context (getConfiguredCabalData ConfCabal.cSrcs)
+    objs <- mapM (objectPath context) srcs
+    return $ if way context == threaded
+        then objs
+        else filter ((`notElem` ["Evac_thr", "Scav_thr"]) . takeBaseName) objs
```

Comment:
As far as I know we no longer need any special hacks for `Evac_thr` and `Scav_thr`. Why is this back?

Reply:
@angerman ?

Reply:
The `cSrcs` from the cabal file will still include the `Evac_thr` and `Scav_thr` files. What we do not need to do anymore is the stupid copying of `Evac` to `Evac_thr` and `Scav` to `Scav_thr`.  However we do not want to include the `_thr` versions when we are not building a threaded runtime.

*if* the `rts.cabal` file learns about `-fthreaded`, this could be flag-guarded in the cabal file.  Until then,
this is required to be guarded here.

Note: while we do have the rts.cabal file now, which allows us to pull info about the rts from the cabal description, Cabal is not able to actually build the rts. That's also why hadrian still does all this assembling of the rts. Cabal has no concept of building multiple library flavours in the same package. So far it only has the *understanding* of this being possible without any idea how to construct one itself. That means: `cabal copy/register/install` will work, but `cabal build` will.

Reply:
> if the rts.cabal file learns about -fthreaded, this could be flag-guarded in the cabal file. Until then,
this is required to be guarded here.

That would be a little bit annoying though. Because if we're totally transparent, that we'll need to do a bunch of cascading `if flag(...)` to successfully expose all the relevant flavours that we shall be building. For example we'd need both the `threaded` and `profiling` flags enabled if we want `_thr_p` and would therefore need something along the lines of:

```
if flag(threaded)
  extra-library-flavours: _thr
if flag(profiling)
  extra-library-flavours: _p
  if flag(threaded)
    extra-library-flavours: _thr_p
```

no?

Reply:
@angerman I see, thank you!